pygdf/columnops.py

```diff
@@ -153,33 +148,87 @@ def as_column(arbitrary):
-                npdata = np.asarray(padata.buffers()[1]).view(np.float64)
-                data = as_column(npdata)
-                mask = np.asarray(padata.buffers()[0])
-                data = data.set_mask(mask)
+            data = as_column(cuda.to_device(arbitrary))
+
+    elif isinstance(arbitrary, pa.Array):
+        if isinstance(arbitrary, pa.StringArray):
+            raise NotImplementedError("Strings are not yet supported")
+        elif isinstance(arbitrary, pa.NullArray):
+            pamask = Buffer(np.empty(0, dtype='int8'))
+            padata = Buffer(
+                np.empty(0, dtype=arbitrary.type.to_pandas_dtype())
+            )
+            data = numerical.NumericalColumn(
+                data=padata,
+                mask=pamask,
+                null_count=0,
+                dtype=np.dtype(arbitrary.type.to_pandas_dtype())
+            )
+        elif isinstance(arbitrary, pa.DictionaryArray):
+            if arbitrary.buffers()[0]:
+                pamask = Buffer(np.array(arbitrary.buffers()[0]))
+            else:
+                pamask = None
+            padata = Buffer(np.array(arbitrary.buffers()[1]).view(
+                arbitrary.indices.type.to_pandas_dtype()
+            ))
+            data = categorical.CategoricalColumn(
+                data=padata,
+                mask=pamask,
+                null_count=arbitrary.null_count,
+                categories=arbitrary.dictionary.to_pylist(),
+                ordered=arbitrary.type.ordered,
+                dtype="category"  # What's the correct way to specify this?
+            )
+        elif isinstance(arbitrary, pa.TimestampArray):
+            arbitrary = arbitrary.cast(pa.timestamp('ms'))
+            if arbitrary.buffers()[0]:
+                pamask = Buffer(np.array(arbitrary.buffers()[0]))
+            else:
+                pamask = None
+            padata = Buffer(np.array(arbitrary.buffers()[1]).view(
+                np.dtype('M8[ms]')
+            ))
+            data = datetime.DatetimeColumn(
+                data=padata,
+                mask=pamask,
+                null_count=arbitrary.null_count,
+                dtype=np.dtype('M8[ms]')
+            )
+        else:
+            if arbitrary.buffers()[0]:
+                pamask = Buffer(np.array(arbitrary.buffers()[0]))
+            else:
+                pamask = None
+            padata = Buffer(np.array(arbitrary.buffers()[1]).view(
+                np.dtype(arbitrary.type.to_pandas_dtype())
+            ))
+            data = numerical.NumericalColumn(
+                data=padata,
+                mask=pamask,
+                null_count=arbitrary.null_count,
+                dtype=np.dtype(arbitrary.type.to_pandas_dtype())
+            )
+
+    elif isinstance(arbitrary, pd.Categorical):
+        # Necessary because of ARROW-3324
+        from .categorical import pandas_categorical_as_column
+        data = pandas_categorical_as_column(arbitrary)
```

Comment:
This could be updated to go through pyarrow but would need to be wrapped in a try except for if there's NaNs in the Pandas Series/Categorical as PyArrow does not yet handle that condition. Thoughts?

Reply:
Presumably going directly to pyarrow is faster than going through pandas? It seems as though going via pandas may cause a double copy of the data. Is this correct?